### PR TITLE
[react-native] added missed onChange method to SwitchProps

### PR DESCRIPTION
**Pull request by mr-africa**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://reactnative.dev/docs/switch#onchange

`onChange` method doesn't exists in type definition

### CODE REVIEW

File: types/react-native/index.d.ts
Comment (ca057) [inline types/react-native/index.d.ts:8640]: Two remarks based on https://github.com/facebook/react-native/blob/master/Libraries/Components/Switch/Switch.js#L79:
* Can we make the `onChange`-prop not only accept `undefined` but also `null`, as it is a maybe-type in the source?
* Can we type the `SwitchChangeEvent` as well?

And do you mind adding some tests?

**Review by ca057 — CHANGES_REQUESTED** [review]

File: types/react-native/index.d.ts
Comment (mr-africa) [reply types/react-native/index.d.ts]: * I added some tests
* I added `null` accepted type

> Can we type the SwitchChangeEvent as well?

unfortunately react-native doesn't export `SwitchChangeEvent`. So I can to add it to declaration types file, but I can't import this event in my application (or I don't know how to do it) see https://github.com/facebook/react-native/blob/master/Libraries/Components/Switch/Switch.js#L27

Good solution will be:

* add `export SwitchChangeEvent` to react-native repo
* open pull request
* wait to merge
* type SwitchChangeEvent here

But it might spend a lot of time

Comment (ca057) [reply types/react-native/index.d.ts]: Hey :)

Sorry, I think my first comment was a bit misleading, as the maybe type from flow (https://flow.org/en/docs/types/maybe/) applies to the whole prop (it can be `null`, `undefined` or the callback):
```diff
- onChange?: (event?: any) => Promise<void> | void | null;
+ onChange?: ((event?: any) => Promise<void> | void) null;
```

In regards to the event: You can simply declare the event as interface in the file and use it here, similar additions were done for other events: https://github.com/DefinitelyTyped/DefinitelyTyped/blob/777174d7912b0dbe86fe0081f00c196d626956e8/types/react-native/index.d.ts#L410

Comment (mr-africa) [reply types/react-native/index.d.ts]: @ca057 thanks for your advices. I've fixed everything. Look it again please

**Review by ca057 — APPROVED** [review]

File: types/react-native/index.d.ts
Comment (ca057) [reply types/react-native/index.d.ts]: Looks good to me!

File: types/react-native/index.d.ts
Comment (amcasey) [inline types/react-native/index.d.ts:8644]: I'm not familiar with this API, but it looks like `onValueChange`, below, can't be promise-returning or `null`.  Is the mismatch intentional?

Comment (amcasey) [reply types/react-native/index.d.ts:8644]: Incidentally, you should be able to assign a promise-returning function to a void-returning function type, so I would expect the tests to pass without the `Promise<void> |`.

Comment (amcasey) [reply types/react-native/index.d.ts:8644]: As for `null`, I'll let you decide, but the two should probably be consistent.

**Review by amcasey — CHANGES_REQUESTED** [review]
Questions

File: types/react-native/index.d.ts
Comment (mr-africa) [reply types/react-native/index.d.ts:8644]: I added `Promise` return type as used in source code  https://github.com/facebook/react-native/blob/master/Libraries/Components/Switch/Switch.js#L79 

I checked locally, all tests are passed without `Promise<void> |`. So I will remove it if you think it's for the best

As for `null`, @ca057 said me to add `null` check in the conversation above

Comment (amcasey) [reply types/react-native/index.d.ts:8644]: I don't feel strongly about either change, but I think the two signatures should agree.

Comment (ca057) [reply types/react-native/index.d.ts:8644]: > I think the two signatures should agree.
Good point. I think we should follow the typings in the React Native codebase, as out of my perspective this should be the source of truth which is the most likely to be correct. I would therefore propose to adapt the typing of `onValueChange` as well to properly reflect the maybe-type and the possible return values. What do you think?

Maybe @alloy can comment how these adaptions were done previously in the typings?

Comment (mr-africa) [reply types/react-native/index.d.ts:8644]: @ca057 I totally agree with you that react native codebase should be source of truth for typing.

Should I change `onValueChange ` method signature as you described above? Or will we be waiting for @alloy answer?

Comment (ca057) [reply types/react-native/index.d.ts:8644]: I would align it and fix the `onValueChange` as well according to the types in the React Native repo.

Comment (mr-africa) [reply types/react-native/index.d.ts:8644]: @ca057  `onValueChange` method has been updated

Comment (mr-africa) [reply types/react-native/index.d.ts:8644]: Ci has found type error in react-dialog lib after my changes in method `onValueChange`, should I change this lib types definition too? I'm afraid  it might will broken something else.

**Review by ca057 — APPROVED** [review]

File: types/react-native/index.d.ts
Comment (ca057) [reply types/react-native/index.d.ts:8644]: I think we should fix the types there too. Can you do it or shall I help to get this type update done? :)

**Review by ca057 — APPROVED** [review]

File: types/react-native/index.d.ts
Comment (TheSavior) [reply types/react-native/index.d.ts:8610]: Woah, this type is weird. I have no idea why it is Promise<void> in the main repo. It should just be `mixed` in Flow. It seems reasonable to align with Flow in this PR, but I think the upstream code is wrong too.

Comment (mr-africa) [reply types/react-native/index.d.ts:8610]: @amcasey If you don't have any more questions, could you resolve your "changes requested" status please

**Comment by mr-africa** [comment]
Ready to merge

**Comment by NickGerleman** [comment]
Hmm, this is showing breaks when integrating into our repo.

https://github.com/microsoft/react-native-windows/pull/7905

The specific issue is around the change to `onValueChange`, which marks the value parameter in the callback potentially undefined, breaking anyone who assumes it have a value and has `strict` or `strictNullChecks` enabled.

Was this change because the callback can actually be called without a value, or was it to allow callbacks which do not use the value? The latter should just work.

File: types/react-native/index.d.ts
Comment (amcasey) [reply types/react-native/index.d.ts:8610]: Sorry about that.  Looks like you managed to merge anyway though. :smile:

**Comment by ca057** [comment]
> Hmm, this is showing breaks when integrating into our repo.
> 
> [microsoft/react-native-windows#7905](https://github.com/microsoft/react-native-windows/pull/7905)
> 
> The specific issue is around the change to `onValueChange`, which marks the value parameter in the callback potentially undefined, causing type-checking errors for anyone who assumes it have a value and has `strict` or `strictNullChecks` enabled.
> 
> Was this change because the callback can actually be called without a value, or was it to allow callbacks which do not use the value?

Uff yeah, I think this slipped my eyes when reviewing... We were mostly concerned to align the typings of the two callbacks (`onChange` and `onValueChange`) and their return values to properly reflect the maybe-type etc. from the source typings (https://github.com/facebook/react-native/blob/master/Libraries/Components/Switch/Switch.js#L79) - see also the discussion above. The argument not being optional slipped my eyes and we should fix it (making the argument non-optional). Will you provide a fix or shall I?

**Comment by NickGerleman** [comment]
@ca057 would definitely appreciate a fix if you have a free moment.

**Pull request merged**
